metrics/framework/Metrics.hpp

```diff
@@ -0,0 +1,305 @@
+
+
+/*
+ * IMetric
+ *
+ * Interface defining a metric
+ */
+interface IMetric
+{
+    /*
+     * Returns the metric name
+     */
+    virtual std::string getName() const = 0;
+
+    /*
+     * Returns metric description
+     */
+    virtual std::string getDescription() const = 0;
+
+    /*
+     * Returns the metric type.
+     */
+    virtual MetricType getMetricType() const = 0;
+
+    /*
+     * Get report values
+     */
+    virtual void getReportValues(MeasurementVector &mv) const = 0;
+};
+
+
+/*
+* Concrete base class implementation of the IMetric interface. All metrics inherit
+* from this class.
+*/
+class METRICS_API Metric : public IMetric
+{
+    public:
+        virtual ~Metric() = default;
+        std::string getName() const override { return name; }
+        std::string getDescription() const override { return description; }
+        MetricType getMetricType() const override { return metricType; }
+
+    protected:
+        // No one should be able to create one of these
+        Metric(const char *_name, const char *_desc, MetricType _metricType) :
+                name{_name},
+                description{_desc},
+                metricType{_metricType} { }
+
+    protected:
+        std::string name;
+        std::string description;
+        MetricType metricType;
+};
+
+class METRICS_API CounterMetric : public Metric
+{
+    public:
+
+        CounterMetric(const char *name, const char *description) :
+                Metric{name, description, MetricType::METRICS_GAUGE}  { }
+        ~CounterMetric() override = default;
+        void inc(uint32_t val)
+        {
+            count.fetch_add(val);
+        }
+
+        void getReportValues(MeasurementVector &mv) const override;
+
+
+protected:
+
+    std::atomic<uint32_t> count{0};
+};
+
+
+/*
+ * Metric used to track the current state of some internal measurement.
+ */
+class METRICS_API GaugeMetric : public Metric {
+    public:
+        GaugeMetric(const char *name, const char *description) :
+                Metric{name, description, MetricType::METRICS_GAUGE}  { }
+
+        /*
+         * Update the value as indicated
+         */
+        void add(float val)
+        {
+            auto current = gaugeValue.load();
+            while (!gaugeValue.compare_exchange_weak(current, current + val));
+        }
+        //void inc(float inc) { value += inc; }
+
+        /*
+         * Decrement the value
+         */
+        //void dec(float dec) { value -= dec; }
+
+        /*
+         * Set the value
+         */
+        void set(float val) { gaugeValue = val; }
+
+        /*
+         * Read the current value
+         */
+        float getValue() const { return gaugeValue; }
+
+        /*
+         * Get current gauge value
+         */
+        void getReportValues(MeasurementVector &mv) const override;
+
+
+    protected:
+        std::atomic<float> gaugeValue{0};
+};
+
+
+class MetricsReporter;
+interface IMetricSink
```

Comment:
Should either be derived from IInterface, or have a virtual destructor, otherwise it can't be cleaned up.

Reply:
Not sure I understand. The MetricSink class which provides a base class for all sinks to inherit, implements the IMetricSink interface and provides a virtual destructor.

Reply:
Your factory function returns an IMetricSink.  That means you cannot cleanly dispose of the object returned from the function.